### PR TITLE
Iox #27 use cxx expected for client and server send method

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

This PR changes the client and server `send` method to return a `cxx::expected` to indicate a send error.

The tests for the untyped C++ and C API are adjusted. The tests for the typed C++ API will be done with #1140

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Relates to #27

### CODE REVIEW

File: iceoryx_posh/include/iceoryx_posh/internal/popo/rpc_interface.hpp
Comment (elBoberido) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/rpc_interface.hpp:36]: The tests for the typed API are not done yet and the example is not in master, so I cannot easily test the SFINEA which is necessary. I will change this with the test for the typed API

Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/rpc_interface.hpp]: Wasn't that difficult after all

**Review by mossmaurice — COMMENTED** [review]
LGTM in general, few comments.

File: iceoryx_binding_c/test/moduletests/test_client.cpp
Comment (elfenpiff) [inline iceoryx_binding_c/test/moduletests/test_client.cpp:294]: Could you please add one more test where `iox_client_send` fails. Just to verify that you are not always returning `SUCCESS`.

File: iceoryx_binding_c/test/moduletests/test_server.cpp
Comment (elfenpiff) [inline iceoryx_binding_c/test/moduletests/test_server.cpp:468]: Could you please add at least another test where `iox_server_send` fails so that we can verify that the return value is not always `SUCCESS`.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_distributor.inl
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_distributor.inl:136]: Does the publisher/server has the ability to know how many subscriber/clients they have? I am just wondering what value the return value has and what the intention behind it is.

For my point of view it would maybe make more sense to state to how many queues it was not delivered or simpler, if all queues have received the chunk or not.
But this is me just asking...

This question goes for all layers and levels. Maybe it makes more sense to have a bool which states just if this was delivered to all queues.

File: iceoryx_binding_c/test/moduletests/test_cpp2c_enum_translation.cpp
Comment (elBoberido) [reply iceoryx_binding_c/test/moduletests/test_cpp2c_enum_translation.cpp:217]: Good point. This is used all over the place in the C-Binding. I'll check if it is easy to change it everywhere and will create an issue for this if not

File: iceoryx_posh/include/iceoryx_posh/internal/popo/ports/client_port_user.hpp
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/ports/client_port_user.hpp:36]: In the C-binding there is a `SUCCESS` value and an `UNDEFINED_ERROR`, the first one is `0`, the second one `1`. When this enum starts at 2, the remaining underlying values are equal. I thought it would be a good idea but can also revert this

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_distributor.inl
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_distributor.inl:136]: I used the integer since it give more information than a boolean. We could also think about returning a `(int32_t, int32_t)` tuple with the delivered and skipped queues. If you don't mind, I would create a separate issue for this since it is not public API and for the request response the current implementation with the counter is sufficient

File: iceoryx_posh/source/popo/ports/server_port_user.cpp
Comment (elBoberido) [reply iceoryx_posh/source/popo/ports/server_port_user.cpp:139]: Hahaha naming is hard. I let this sink in while I fix the other stuff and also look at the error handler enums and try to be consistent with these. I don't mind to change it.

File: iceoryx_posh/source/popo/ports/client_port_user.cpp
Comment (elBoberido) [reply iceoryx_posh/source/popo/ports/client_port_user.cpp:90]: Contrary to pub-sub, a client will wait for a response when the request was sent, e.g. with a blocking `WaitSet`. When the `send` failed, one does not need to wait.

File: iceoryx_posh/source/popo/ports/server_port_user.cpp
Comment (elBoberido) [reply iceoryx_posh/source/popo/ports/server_port_user.cpp:139]: @mossmaurice we don't have an error handler enum value with offer and in our codebase we also don't have `OFFERING` -> `NOT_OFFERED`, which matches nicely with `isOffered`

File: iceoryx_binding_c/test/moduletests/test_cpp2c_enum_translation.cpp
Comment (elBoberido) [reply iceoryx_binding_c/test/moduletests/test_cpp2c_enum_translation.cpp]: I changed it for cpp2c enum tests. Let's see if the CI builds on all platform, then I'll change it for c2cpp

File: iceoryx_posh/include/iceoryx_posh/internal/popo/ports/server_port_user.hpp
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/ports/server_port_user.hpp:72]: Is this somewhere tested? Not that someone is relying on this and we break this by accident.

**Review by elfenpiff — COMMENTED** [review]
I have just one more question, after that we can merge -> see comment.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/ports/server_port_user.hpp
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/ports/server_port_user.hpp:72]: It is not really necessary and a micro optimization. @mossmaurice also asked about this. Maybe it is better to remove it for now and decide later on if we want to go this route for all C binding enums

**Review by elfenpiff — APPROVED** [review]

File: iceoryx_binding_c/test/moduletests/test_c2cpp_enum_translation.cpp
Comment (mossmaurice) [inline iceoryx_binding_c/test/moduletests/test_c2cpp_enum_translation.cpp:53]: :heart:

File: iceoryx_binding_c/test/test.hpp
Comment (mossmaurice) [inline iceoryx_binding_c/test/test.hpp:33]: Are you sure this will trigger a warning? The cast should be fine?

**Review by mossmaurice — APPROVED** [review]

File: iceoryx_binding_c/test/test.hpp
Comment (elBoberido) [reply iceoryx_binding_c/test/test.hpp:33]: It triggers the warning from the c2cpp conversion

**Review by mossmaurice — APPROVED** [review]

**Review by elfenpiff — APPROVED** [review]

**Pull request merged**
